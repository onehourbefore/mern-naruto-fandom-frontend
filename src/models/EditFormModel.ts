

export enum TasksEnum {
    CREATE= 'create',
    UPDATE = 'update',
    IDLE = 'idle'
}

export interface EditFormState {
    postID: string
    task: TasksEnum
}