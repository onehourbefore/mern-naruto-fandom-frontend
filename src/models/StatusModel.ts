export enum StatusModel {
    LOADING = 'loading',
    ERROR = 'error',
    SUCCESS = 'success',
    IDLE = 'idle'
}