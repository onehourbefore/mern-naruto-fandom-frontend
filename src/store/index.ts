import { configureStore, combineReducers } from '@reduxjs/toolkit';
import { useDispatch, TypedUseSelectorHook, useSelector } from 'react-redux';

import posts from './slices/postsSlice';
import filter from './slices/filterSlice';
import onePost from './slices/onePostSlice';
import registration from './slices/registrationSlice';
import authorization from './slices/authorizationSlice';
import experts from './slices/expertSlice';
import comments from './slices/commentSlice';
import editForm from './slices/editFormSlice';

export const rootReducer = combineReducers ({
    posts,
    filter,
    onePost,
    registration,
    authorization,
    experts,
    comments,
    editForm,
});

export const store = configureStore ({
    reducer: rootReducer,
    middleware: getDefaultMiddleware => getDefaultMiddleware (),
    devTools: process.env.NODE_ENV !== 'production'
});

export type RootState = ReturnType <typeof store.getState>;
export const useAppDispatch: () => typeof store.dispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;