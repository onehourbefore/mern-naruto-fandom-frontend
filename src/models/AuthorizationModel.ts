import { StatusModel } from "./StatusModel"

export type UserDataType = {
    id: string,
    email: string,
    name: string,
    role: string,
    avatar: string,
    created: number,
    liked: string [] | []
}

export type ResponseType = {
    accessToken: string,
    refreshToken: string,
    user: UserDataType
}

export interface AuthorizationState {
    user: UserDataType,
    status: StatusModel,
    deleteProfileStatus: StatusModel
}