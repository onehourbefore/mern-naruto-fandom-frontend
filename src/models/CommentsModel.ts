import { StatusModel } from "./StatusModel"

export type CommentType = {
    _id: string,
    postID: string,
    date: string,
    name: string,
    email: string,
    body: string,
    avatar: string,
}

export interface CommentsState {
    comments: [] | CommentType [],
    status: StatusModel,
}

export type CommentProps = {
    data: {
        _id: string,
        name: string,
        email: string,
        body: string,
        avatar: string,
        date: string,
    },
    postID: string | undefined,
    commentIndex: number,
}

export type CommentFormProps = {
    setVisible: (arg: boolean) => void,
}

export type CommentFromSidebarProps = {
    comment: CommentType,
}