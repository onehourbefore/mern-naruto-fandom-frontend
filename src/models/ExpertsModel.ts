import { StatusModel } from "./StatusModel"

export type ExpertType = {
    avatar: string,
    email: string,
    name: string,
    created: number
}


export interface ExpertsState {
    experts: [] | ExpertType [],
    status: StatusModel
}