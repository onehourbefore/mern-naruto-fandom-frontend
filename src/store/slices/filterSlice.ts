import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import $api from '../../http';
import { PostType } from '../../models/PostModel';
import { FilterState } from '../../models/FilterModel';
import { StatusModel } from '../../models/StatusModel';
import { FilterThunks } from './thunks/FilterThunks';

const filterThunks = new FilterThunks();

const initialState: FilterState = {
    list: [],
    tags: [],
    fetchStatus: StatusModel.IDLE,
    searchStatus: StatusModel.IDLE,
}

export const findPosts = createAsyncThunk(
    'searchPost',
    async(q: {type: '_query' | '_tag', payload: string}, { rejectWithValue }) => {
        try {
            return await filterThunks.findPostsCallback(q);
        } catch(e: any) {
            return rejectWithValue(e.message);
        }
    }
)

export const getAllTags = createAsyncThunk(
    'getAllTags',
    async(_, { rejectWithValue }) => {
        try {
            return await filterThunks.getAllTagsCallback();
        } catch(e: any) {
            return rejectWithValue(e.message);
        }
    }
)

export const filterSlice = createSlice({
    name: 'filter',
    initialState,
    reducers: {
        clearFiltered: (state) => {
            state.list = [];
        },
        clearAllTags: (state) => {
            state.tags = [];
        },
        clearSearchStatus: (state) => {
            state.searchStatus = StatusModel.IDLE;
        }
    },
    extraReducers: builder => {
        builder
        .addCase(findPosts.pending, (state) => {
            state.searchStatus = StatusModel.LOADING;
        })
        .addCase(findPosts.fulfilled, (state, action: PayloadAction <PostType[]>) => {
            if (action.payload.length !== 0) {
                state.list = action.payload;
                state.searchStatus = StatusModel.SUCCESS;
            } else {
                state.searchStatus = StatusModel.ERROR;
            }
        })
        .addCase(findPosts.rejected, (state) => {
            state.searchStatus = StatusModel.ERROR;
        })


        .addCase(getAllTags.pending, (state) => {
            state.fetchStatus = StatusModel.LOADING;
        })
        .addCase(getAllTags.fulfilled, (state, action: PayloadAction <string []>) => {
            state.tags = action.payload;
            state.fetchStatus = StatusModel.SUCCESS;
        })
        .addCase(getAllTags.rejected, (state) => {
            state.fetchStatus = StatusModel.ERROR;
        })

        .addDefaultCase(() => {})
    }
})

const { actions, reducer } = filterSlice;
export default reducer;
export const { clearFiltered, clearAllTags, clearSearchStatus } = actions;