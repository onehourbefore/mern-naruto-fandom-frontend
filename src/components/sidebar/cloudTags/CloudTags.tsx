import React from 'react';
import { useAppDispatch, useAppSelector } from '../../../store';
import { getAllTags } from '../../../store/slices/filterSlice';
import { StatusModel } from '../../../models/StatusModel';
import { getFilter, getPostsAll } from '../../../selectors';
import Spinner from '../../spinner/Spinner';
import TitleOfCloudTags from './components/TitleOfCloudTags';
import Tags from './components/Tags';

import cl from './CloudTags.module.scss';


const CloudTags: React.FC = (): JSX.Element => {
    const dispatch = useAppDispatch();
    const { posts } = useAppSelector(getPostsAll);
    const { fetchStatus } = useAppSelector(getFilter);
    const [selectedTag, setSelectedTag] = React.useState('');

    React.useEffect(() => {
        if (posts.length)  {
            dispatch(getAllTags());
        }
    }, [posts]);

    return (
        <div className={cl.root}>
            <TitleOfCloudTags
                setSelectedTag={setSelectedTag}
            />
            <div className={cl.root__body}>
                {fetchStatus === StatusModel.LOADING
                ? <Spinner mt='0px' />
                : <Tags
                    selectedTag={selectedTag}
                    setSelectedTag={setSelectedTag}
                />}
            </div>
        </div>
    )
};

export default CloudTags;