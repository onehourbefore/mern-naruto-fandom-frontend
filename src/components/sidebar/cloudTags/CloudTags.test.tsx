import { render, screen } from '@testing-library/react';
import * as reduxHooks from '../../../store';
import * as actions from '../../../store/slices/filterSlice';
import { StatusModel } from '../../../models/StatusModel';
import CloudTags from './CloudTags';


const mockedUseAppDispatch = jest.spyOn(reduxHooks, 'useAppDispatch');
const mockedUseAppSelector = jest.spyOn(reduxHooks, 'useAppSelector');
const mockedGetAllTagsAction = jest.spyOn(actions, 'getAllTags');

const returnState = {
    posts: [],
    fetchStatus: StatusModel.LOADING,
    list: [],
};

describe('CloudTags', () => {
    it('status loading', async () => {
        mockedUseAppDispatch.mockReturnValue(jest.fn());
        mockedUseAppSelector.mockReturnValue(returnState);
        const component = render(
            <CloudTags />
        );
        const spinner = await screen.findByTestId('spinner');
        expect(spinner).toBeInTheDocument();

        expect(component).toMatchSnapshot();
    });

    it('status success (posts.length = 0)', async () => {
        mockedUseAppDispatch.mockReturnValue(jest.fn());
        mockedUseAppSelector.mockReturnValue({
            ...returnState,
            fetchStatus: StatusModel.SUCCESS,
            tags: ['first', 'second', 'third'],
        });
        const component = render(
            <CloudTags />
        );

        expect(screen.getByText('first')).toBeInTheDocument();
        expect(screen.getByText('second')).toBeInTheDocument();
        expect(screen.getByText('third')).toBeInTheDocument();
        expect(mockedGetAllTagsAction).toHaveBeenCalledTimes(0);

        expect(component).toMatchSnapshot();
    });

    it('status success (posts.length > 0)', async () => {
        mockedUseAppDispatch.mockReturnValue(jest.fn());
        mockedUseAppSelector.mockReturnValue({
            ...returnState,
            fetchStatus: StatusModel.SUCCESS,
            posts: [{}, {}, {}],
            tags: ['first', 'second', 'third'],
        });
        const component = render(
            <CloudTags />
        );

        expect(screen.getByText('first')).toBeInTheDocument();
        expect(screen.getByText('second')).toBeInTheDocument();
        expect(screen.getByText('third')).toBeInTheDocument();
        expect(mockedGetAllTagsAction).toHaveBeenCalledTimes(1);

        expect(component).toMatchSnapshot();
    });
});