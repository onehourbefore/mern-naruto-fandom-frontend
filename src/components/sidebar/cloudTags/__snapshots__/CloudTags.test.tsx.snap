// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudTags status loading 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
      >
        <h3
          class="root__title"
        >
          ОБЛАКО ТЕГОВ
        </h3>
        <div
          class="root__body"
        >
          <div
            class="spinner"
            data-testid="spinner"
            style="margin-top: 0px;"
          >
            <div
              class="ldsRing"
            >
              <div />
              <div />
              <div />
              <div />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
    >
      <h3
        class="root__title"
      >
        ОБЛАКО ТЕГОВ
      </h3>
      <div
        class="root__body"
      >
        <div
          class="spinner"
          data-testid="spinner"
          style="margin-top: 0px;"
        >
          <div
            class="ldsRing"
          >
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CloudTags status success (posts.length = 0) 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
      >
        <h3
          class="root__title"
        >
          ОБЛАКО ТЕГОВ
        </h3>
        <div
          class="root__body"
        >
          <span
            class="root__tag"
          >
            first
          </span>
          <span
            class="root__tag"
          >
            second
          </span>
          <span
            class="root__tag"
          >
            third
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
    >
      <h3
        class="root__title"
      >
        ОБЛАКО ТЕГОВ
      </h3>
      <div
        class="root__body"
      >
        <span
          class="root__tag"
        >
          first
        </span>
        <span
          class="root__tag"
        >
          second
        </span>
        <span
          class="root__tag"
        >
          third
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CloudTags status success (posts.length > 0) 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
      >
        <h3
          class="root__title"
        >
          ОБЛАКО ТЕГОВ
        </h3>
        <div
          class="root__body"
        >
          <span
            class="root__tag"
          >
            first
          </span>
          <span
            class="root__tag"
          >
            second
          </span>
          <span
            class="root__tag"
          >
            third
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
    >
      <h3
        class="root__title"
      >
        ОБЛАКО ТЕГОВ
      </h3>
      <div
        class="root__body"
      >
        <span
          class="root__tag"
        >
          first
        </span>
        <span
          class="root__tag"
        >
          second
        </span>
        <span
          class="root__tag"
        >
          third
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
